packages/ui/src/contexts/AppearanceContext.tsx

```diff
@@ -0,0 +1,421 @@
+
+  const ctxValue = useDeepEqualMemo(() => {
+    const theme = props.theme ?? globalTheme;
+    const themelessAppearance = mergeAppearance(globalThemelessAppearance, props.appearance);
+    const parsedAppearance = applyTheme(theme, themelessAppearance);
+
+    return { value: { parsedAppearance, themelessAppearance, theme: props.theme ?? globalTheme } };
+  }, [props.appearance, props.theme, globalAppearance, globalThemelessAppearance, globalTheme]);
+
+  return <AppearanceContext.Provider value={ctxValue}>{props.children}</AppearanceContext.Provider>;
+};
+
+export { AppearanceProvider, useAppearance };
+
+if (import.meta.vitest) {
+  const { it, expect, describe } = import.meta.vitest;
+
+  describe('mergeAppearance', () => {
+    it('retains keys from both appearances', () => {
+      const a = { elements: { alert__warning: 'cl-test-class-one' } };
+      const b = { elements: { alertIcon: 'cl-test-class-two' } };
+      expect(mergeAppearance(a, b)).toStrictEqual({
+        layout: {},
+        elements: {
+          alert__warning: 'cl-test-class-one',
+          alertIcon: 'cl-test-class-two',
+        },
+      });
+    });
+
+    it('retains the theme prop', () => {
+      const a = { theme: fullTheme, elements: { alert__warning: 'cl-test-class-one' } };
+      const b = {
+        elements: { alertIcon: 'cl-test-class-two' },
+      };
+      expect(mergeAppearance(a, b)).toStrictEqual({
+        layout: {},
+        theme: a.theme,
+        elements: {
+          alert__warning: 'cl-test-class-one',
+          alertIcon: 'cl-test-class-two',
+        },
+      });
+    });
+
+    it('overrides the theme prop', () => {
+      const a = { theme: fullTheme, elements: { alert__warning: 'cl-test-class-one' } };
+      const b = {
+        theme: { ...fullTheme, alertIcon: { descriptor: 'test', className: 'cl-test-class', style: {} } },
+        elements: { alertIcon: 'cl-test-class-two' },
+      };
+      expect(mergeAppearance(a, b)).toStrictEqual({
+        layout: {},
+        theme: b.theme,
+        elements: {
+          alert__warning: 'cl-test-class-one',
+          alertIcon: 'cl-test-class-two',
+        },
+      });
+    });
+
+    it('merges string values for the same element', () => {
+      const a = { elements: { alert__warning: 'cl-test-class-one' } };
+      const b = { elements: { alert__warning: 'cl-test-class-two' } };
+      expect(mergeAppearance(a, b)).toStrictEqual({
+        layout: {},
+        elements: {
+          alert__warning: 'cl-test-class-one cl-test-class-two',
+        },
+      });
+    });
+
+    it('merges object values for the same element', () => {
+      const a = { elements: { alert__warning: { background: 'tomato' } } };
+      const b = { elements: { alert__warning: { color: 'red' } } };
+      expect(mergeAppearance(a, b)).toStrictEqual({
+        layout: {},
+        elements: {
+          alert__warning: { color: 'red', background: 'tomato' },
+        },
+      });
+    });
```

Comment:
should we add a test where to style objects contain a color key? Does the one farthest down the tree take precedence?

Reply:
good catch! https://github.com/clerk/javascript/pull/3976/commits/1cdd7ffaec21b4499fecc6e2232447ec02f10ca5